File: LeetCode/1463. Cherry Pickup II/Solution.py
class Solution:
    def cherryPickup(self, grid: List[List[int]]) -> int:
        row = len(grid)
        col = len(grid[0])

        dp = [[ [0 for col2 in range(col)] for col1 in range(col)] for row in range(2)]

        dirs = [-1, 0, 1]

        u = 0

        for r in range(row - 1, -1, -1):
            for c1 in range(0, col, 1):
                for c2 in range(c1, col, 1):
                    cherry = grid[r][c1]
                    if c1 != c2: cherry += grid[r][c2]

                    dp[1 - u][c1][c2] = cherry

                    if r != row - 1:
                        for dir1 in dirs:
                            for dir2 in dirs:
                                nextC1 = c1 + dir1
                                nextC2 = c2 + dir2

                                if nextC1 < 0 or nextC1 == col: continue
                                if nextC2 < 0 or nextC2 == col: continue

                                if(nextC1 <= nextC2):
                                    dp[1-u][c1][c2] = max(dp[1-u][c1][c2], cherry + dp[u][nextC1][nextC2])
            u = 1 - u
        
        return dp[u][0][col - 1]


        

File: LeetCode/1463. Cherry Pickup II/Solution.cs
public class Solution
{
    private int[] dirs = new int[] { -1, 0, 1 };

    public int CherryPickup(int[][] grid)
    {
        var row = grid.Length;
        var col = grid[0].Length;

        var dp = new int[2, col, col];

        var u = 0;

        for (var r = row - 1; r >= 0; r--)
        {
            for (var c1 = 0; c1 < col; c1++)
                for (var c2 = c1; c2 < col; c2++)
                {
                    if (r == row - 1) dp[1 - u, c1, c2] = c1 != c2 ? (grid[r][c1] + grid[r][c2]) : grid[r][c1];
                    else
                    {
                        var cherry = c1 != c2 ? (grid[r][c1] + grid[r][c2]) : grid[r][c1];

                        dp[1 - u, c1, c2] = cherry;

                        foreach (var dir1 in dirs)
                            foreach (var dir2 in dirs)
                            {
                                var nextc1 = c1 + dir1;
                                var nextc2 = c2 + dir2;
                                if (nextc1 < 0 || nextc1 == col) continue;
                                if (nextc2 < 0 || nextc2 == col) continue;
                                if (nextc1 <= nextc2)
                                {
                                    dp[1 - u, c1, c2] = Math.Max(dp[1 - u, c1, c2], cherry + dp[u, nextc1, nextc2]);
                                }
                            }
                    }
                }

            u = 1 - u;
        }

        return dp[u, 0, col - 1];
    }
}